### DISCUSSION TITLE
Code Generation and S-functions using osqp with nlpsol using SQPmethod

### DISCUSSION THREAD

**Discussion by Nilstan** [casadi/casadi#3241] [Q&A]
I am currently trying to solve an NLP problem using CasADi and OSQP within a Simulink model. I have followed the example provided in the blog post titled CasADi codegen and S-Functions (https://web.casadi.org/blog/s-function/), which served as my baseline. 
However, I encountered an error when running the compiled s-function in simulink.

Here's a summary of my setup and the steps I've taken:
- I utilized the code snippet provided in the CasADi blog post as a starting point and made modifications to incorporate OSQP as the QP solver for the SQP method.
- I successfully compiled the code and generated the S-function without any errors or warnings.

However, when I run the Simulink model with the S-function, I receive the following error message:
_An error occurred while stepping through the simulation and the simulation was terminated.
Caused by: FatalException_

Upon expanding the error message, I found the following additional information:
_This segmentation violation occurred while executing the S-function 's_function' in block 'demo/S-Function'.
A common cause of this segmentation violation is an incorrect input port direct feedthrough setting. Each input port of the S-function that is read (accessed) in mdlOutputs and/or mdlGetTimeOfNextVarHit must specify that it needs its input signal in these routines by setting direct feedthrough for these input ports.
Another cause can be incorrect memory accesses which occur when your code accesses beyond the end of an array. For example, if you access input port 5 and in mdlInitializeSizes specify that you only have 4 input ports.
To debug your C-MEX S-function, you can enable diagnostics by compiling the S-function source with the -g flag, e.g.,
mex -g sfunction_name.c_

I suspect that there might be an issue with the compilation step or a mistake in my implementation. Here are some additional details that might be helpful:
- I followed the instructions provided in the CasADi blog post mentioned earlier.
- I have verified that the OSQP library (osqp.lib) and the corresponding header files are located in the same directory as my source files (s_function.c and f.c).
- I have not made any modifications to the s_function.c file provided in the blog post example.

Here is the code that i am using: 
```
import casadi.*

x = MX.sym('x',2);
y = MX.sym('y');
w = dot(x,y*x);
z = sin(x)+y+w;
solver = nlpsol('solver','sqpmethod',struct('x',y,'f',y^2),struct('qpsol','osqp'));

res = solver('x0',y);
% CasADi function with two inputs and two outputs
f = Function('f',{x,y},{w,z+res.x});

% Use the numeric types from simulink to codegenerate
libPath = fullfile(pwd);
incPath = fullfile(pwd);
cg_options = struct;
cg_options.casadi_real = 'real_T';
cg_options.real_min    = num2str(realmin); % Needed if you code-generate sqpmethod method  
cg_options.casadi_int  = 'int_T';
cg_options.with_header = true;
cg = CodeGenerator('f',cg_options);
cg.add_include('simstruc.h');
cg.add(f);
cg.generate();

disp('CasADi codegen completed: created f.c')

%%
pause(1)

disp('Compiling S function')
mex('-g',...        % verbose mode
    ['-I' incPath],...      % includes
    ['-L' libPath],...      % used lib path
    ['osqp.lib'],...           % osqp lib for linking
    's_function.c',...
    ['f.c']...
    );

%%
open_system('demo')
```

If anyone has encountered a similar issue, I would greatly appreciate your assistance in troubleshooting this problem. Any suggestions, insights, or recommendations on how to resolve this error would be highly valuable.

Thank you in advance for your help!

**Comment by phmandl** [comment]
Have a similar problem; with the working example from [here](https://web.casadi.org/blog/s-function/ ), I only get it to work with qrqp. When trying with osqp, mex can't find the header files. Configuring mex similar to you with

```
libPath = fullfile(pwd); 
incPath = fullfile(pwd);
mex(['-I' incPath],...      % includes
    ['-L' libPath],...      % used lib path
    'osqp.lib', ...    
    's_function.c',...
    'f.c'...
    );
```

I get the following compilation warnings: 

```
C:\...\Z_playground\casadi_sqp_CA\f.c: In function 'f_f6_init_mem':
C:\...\Z_playground\casadi_sqp_CA\f.c:16804:7: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]
   A.i = A_row;
       ^
C:\...\Z_playground\casadi_sqp_CA\f.c:16805:7: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]
   A.p = A_colind;
       ^
C:\...\Z_playground\casadi_sqp_CA\f.c:16811:7: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]
   H.i = H_row;
       ^
C:\...\Z_playground\casadi_sqp_CA\f.c:16812:7: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]
   H.p = H_colind;
       ^
```

I then tried to run the s-function in simulink, which crashes matlab because of a Access violation probalby because of the warning. Any ideas how to fix this?

**Comment by jgillis** [comment]
@Nilstan You seem to have found a regression bug in 3.6 where the order of generated function is wrong.
Please try the following: open `f.c`, locate this snippet of code:
```
casadi_real casadi_fmax(casadi_real x, casadi_real y) {
/* Pre-c99 compatibility */
#if __STDC_VERSION__ < 199901L
  return x>y ? x : y;
#else
  return fmax(x, y);
#endif
}
```
and move it to before the `casadi_max_viol` snippet.

@phmandl You are using CasADi 3.5.5 which also had a separate bug. I recommend switching to the latest and trying the suggested patch if you want osqp.
The following flags may also be helpful:
```
    ['-I' GlobalOptions.getCasadiIncludePath()],...
    ['-L' GlobalOptions.getCasadiPath()],...
    '-losqp',...
```

**Reply by phmandl** [reply]
Tried it with your suggestions and with 3.6.2. OSQP is now found but still throws an error:

```
Error using mex
C:\Users\phman\repos\Z_casadi\libosqp.a(osqp.c.obj):osqp.c:(.text+0x2a): undefined reference to `__imp___acrt_iob_func'
C:\Users\phman\repos\Z_casadi\libosqp.a(util.c.obj):util.c:(.text+0x2a): undefined reference to `__imp___acrt_iob_func'
C:\Users\phman\repos\Z_casadi\libosqp.a(cs.c.obj):cs.c:(.text+0x2a): undefined reference to `__imp___acrt_iob_func'
C:\Users\phman\repos\Z_casadi\libosqp.a(qdldl_interface.c.obj):qdldl_interface.c:(.text+0x17a): undefined reference to
`__imp___acrt_iob_func'
C:\Users\phman\repos\Z_casadi\libosqp.a(auxil.c.obj):auxil.c:(.text+0x2a): undefined reference to `__imp___acrt_iob_func'
C:\Users\phman\repos\Z_casadi\libosqp.a(error.c.obj):error.c:(.text+0x25): more undefined references to `__imp___acrt_iob_func'
follow
collect2.exe: error: ld returned 1 exit status


Error in main_mpc (line 131)
mex('-g',...        % verbose mode
```
